### ISSUE TITLE
Client request only authorized for same repository: unauthorized when changing repo name

### ISSUE THREAD

**Issue by borjamunozf** [oras-project/oras-py#133]
Hello.

We have started using the oras sdk but we got an strange behaviour. This is the following scenario:

**Scenario**
- artifacts pushed to _exampleregistry_ - repository _quefacemos_
- artifacts pushed to same _exampleregistry- repository _quefacemos2_
- authenticate oras client succesfully with access token from _az acr registry_

pseudocode

**Actual behaviour**
- tags_quefacemos = cli.get_tags("exampleregistry/quefacemos") succesful request.
- tags_quefacemos2 = cli.get_tags("exampleregistry/quefacemos2") fails and returns Unauthorized.
- If we add a new line to login again between the get_tags call, it works!
fail

```go
cli = oras.client.OrasClient()
cli.login(username=whatever, password=accessToken)
tags_quefacemos = cli.get_tags("exampleregistry/quefacemos")
tags_quefacemos2 = cli.get_tags("exampleregistry/quefacemos2")
```

ok
```go
cli = oras.client.OrasClient()
cli.login(username=whatever, password=accessToken)
tags_quefacemos = cli.get_tags("exampleregistry/quefacemos")
cli.login(username=whatever, password=accessToken)
tags_quefacemos2 = cli.get_tags("exampleregistry/quefacemos2")
```

Expected behaviour
- Running the same commands from the ORAS Cli works:

```bash
oras repo tags exampleregistry/quefacemos

OK

oras repo tags exampleregistry/quefacemos2

OK
```
Is this expected? Why the authentication seems to disappear or stop to being valid after changing the repository?

**Comment by borjamunozf** [comment]
Ok, more info. It seems that after the first succesfull request, the _headers_ changed from Basic to Bearer:

https://github.com/oras-project/oras-py/blob/cb575abaa0f45ef9f474f0971f579c12ef643170/oras/provider.py#L998

For the second request to quefacemos2 repository, the authHeaderRaw shows info about the Bearer:
- the scope is assigned to the previous repo (quefacemos (in this case cct, does not matter).
- complains about insufficient scope

![imagen](https://github.com/oras-project/oras-py/assets/25910927/9dcb652c-e3a5-43f4-b755-e28e6db626bf)

**Comment by vsoch** [comment]
Which registry is this? The challenge here is that there isn't a standard auth flow, and it's been tweaked over the years to fit niche registry cases.

**Comment by borjamunozf** [comment]
Azure OCI Registry

**Comment by vsoch** [comment]
So azure just wants to keep the basic auth, or is it just missing the scope?

**Comment by borjamunozf** [comment]
Not quite sure, but seems that it does not want neither basic auth. 
If I'm not wrong, this the flow:

- client.login(user, acces_token)
this access token looks like this (omitted info)

```json
{
  "iss": "Azure Container Registry",
  "aud": "rcsdockerregistry.azurecr.io",
  "version": "1.0",
  "grant_type": "refresh_token",
  "permissions": {
    "actions": [
      "read",
      "write",
      "delete",
      "deleted/read",
      "deleted/restore/action"
    ]
  },
````
- the header initially is Basic Auth, but this fails anyway. However, it goes through authenticate_request method, ends getting 200 response in line and updates the header from Basic Auth to Bearer. 

https://github.com/oras-project/oras-py/blob/cb575abaa0f45ef9f474f0971f579c12ef643170/oras/provider.py#L1050-L1060

- After the headers update, the request is succesfully authorized & response is processed.

This only works for the inial request, as I mentioned. Any second request has an slighly different flow:

- The header is already a Bearer Token.
- However, the do_request() and authenticate_request brings 401.

**Comment by vsoch** [comment]
Sounds like we need to keep the basic auth then and this registry does not have support for Bearer? Would that fix the issue?

**Comment by borjamunozf** [comment]
I don't think that the registry is not _supporting_ the Bearer. It does, but only seems to be valid for the each single request, so that's why I need to _login_ each time.

Respecting the basic_auth could be an option to avoid login repeatedly because it looks like having it enforces the correct automatic refresh for the client/registry (transforming from Basic Auth to Bearer).

https://github.com/oras-project/oras-py/blob/cb575abaa0f45ef9f474f0971f579c12ef643170/oras/client.py#L226
- If I set & call the _set_basic_auth function, which is what also the _login_ method does internally, this works.

```python
oras_client.login(hostname=self.server, username=OCI_USER, password=out_token["accessToken"])
oras_client.get_tags(whatever) 
oras_client.set_basic_auth(OCI_USER, self.auth_token)
oras_client.get_tags(whatever_repo2) 
```

I don't know if _persisting_ or respecting the _basic_auth member variable for the Registry class could have side effects for others to be honest:

https://github.com/oras-project/oras-py/blob/cb575abaa0f45ef9f474f0971f579c12ef643170/oras/provider.py#L75C14-L75C26

**Comment by vsoch** [comment]
What we probably need is to separate the auth flow into modules - so you can select a module that has a particular behavior. I'd be open to a PR for that - I won't have time myself imminently soon.

**Comment by vsoch** [comment]
Hi @borjamunozf - I started https://github.com/oras-project/oras-py/pull/134 as an effort to refactor auth into modules. I stripped down the default (the token flow) so I'm interested in feedback about if that works for you now, and if not, what the issue is, and then if there were a "basic auth only" flow (which I added a skeleton for) what you'd like that to look like. This is a fairly big change so it might not go in quickly (we need feedback from folks that use other registries) but I wanted to get us started.
